etcd/etcd.go

```diff
@@ -0,0 +1,206 @@
+}
+
+// New Create a default etcd client
+// It can create a client with default config by env variable.
+// See: env.go
+func New(opts Options) (Client, error) {
+	if opts.Address == "" {
+		opts.Address = EtcdDefaultServerAddr
+	}
+	if opts.Port == 0 {
+		opts.Port = EtcdDefaultPort
+	}
+	if opts.ConfigParser == nil {
+		opts.ConfigParser = defaultConfigParse()
+	}
+	if opts.Prefix == "" {
+		opts.Prefix = EtcdDefaultConfigPrefix
+	}
+	if opts.ServerPathFormat == "" {
+		opts.ServerPathFormat = EtcdDefaultServerPath
+	}
+	if opts.clientPathFormat == "" {
+		opts.clientPathFormat = EtcdDefaultClientPath
+	}
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints: []string{fmt.Sprintf("http://%s:%d", opts.Address, opts.Port)},
+		LogConfig: opts.LoggerConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+	prefixTemplate, err := template.New("prefix").Parse(opts.Prefix)
+	if err != nil {
+		return nil, err
+	}
+	serverNameTemplate, err := template.New("serverName").Parse(opts.ServerPathFormat)
+	if err != nil {
+		return nil, err
+	}
+	clientNameTemplate, err := template.New("clientName").Parse(opts.clientPathFormat)
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		ecli:               etcdClient,
+		parser:             opts.ConfigParser,
+		prefixTemplate:     prefixTemplate,
+		serverPathTemplate: serverNameTemplate,
+		clientPathTemplate: clientNameTemplate,
+	}
+	return c, nil
+}
+
+func (c *client) SetParser(parser ConfigParser) {
+	c.parser = parser
+}
+
+func (c *client) ClientConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error) {
+	return c.configParam(cpc, c.clientPathTemplate, cfs...)
+}
+
+func (c *client) ServerConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error) {
+	return c.configParam(cpc, c.serverPathTemplate, cfs...)
+}
+
+// configParam render config parameters. All the parameters can be customized with CustomFunction.
+// ConfigParam explain:
+//  1. Prefix: KitexConfig by default.
+//  2. ServerPath: {{.ServerServiceName}}/{{.Category}} by default.
+//     ClientPath: {{.ClientServiceName}}/{{.ServerServiceName}}/{{.Category}} by default.
+func (c *client) configParam(cpc *ConfigParamConfig, t *template.Template, cfs ...CustomFunction) (Key, error) {
+	param := Key{}
+
+	var err error
+	param.Path, err = c.render(cpc, t)
+	if err != nil {
+		return param, err
+	}
+	param.Prefix, err = c.render(cpc, c.prefixTemplate)
+	if err != nil {
+		return param, err
+	}
+
+	for _, cf := range cfs {
+		cf(&param)
+	}
+	return param, nil
+}
+
+func (c *client) render(cpc *ConfigParamConfig, t *template.Template) (string, error) {
+	var tpl bytes.Buffer
+	err := t.Execute(&tpl, cpc)
+	if err != nil {
+		return "", err
+	}
+	return tpl.String(), nil
+}
+
+// RegisterConfigCallback register the callback function to etcd client.
+func (c *client) RegisterConfigCallback(ctx context.Context, cancel context.CancelFunc, key string, callback func(string, ConfigParser)) {
```

Comment:
This implementation is OK if there's only one listener for one specific key. What if we have 2 listeners for the same key? Especially when one of them needs to cancel the listening, would the other be affected?